oath/src/main/java/com/yubico/yubikit/oath/Base32.java

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright (C) 2024 Yubico.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.yubico.yubikit.oath;
+
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+
+/**
+ * Base32 implementation of RFC4684, section 6
+ *
+ * @see <a href="https://datatracker.ietf.org/doc/html/rfc4648#section-6">Base 32 Encoding</a>
+ */
+@SuppressWarnings("SpellCheckingInspection")
+public class Base32 {
+    private static final char[] ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567=".toCharArray();
+
+    private static final char PADDING = '=';
+
+    private static final String[] ENCODE_PADDING = new String[]{
+            "", "======", "====", "===", "="
+    };
+
+    public static String encode(byte[] data) {
+        StringBuilder buf = new StringBuilder();
+
+        int len = data.length;
+        for (int i = 0; i < len; i += 5) {
+            int b0 = Byte.toUnsignedInt(data[i]);
+            int b1 = len > i + 1 ? Byte.toUnsignedInt(data[i + 1]) : 0;
+            int b2 = len > i + 2 ? Byte.toUnsignedInt(data[i + 2]) : 0;
+            int b3 = len > i + 3 ? Byte.toUnsignedInt(data[i + 3]) : 0;
+            int b4 = len > i + 4 ? Byte.toUnsignedInt(data[i + 4]) : 0;
+
+            buf.append(ALPHABET[b0 >> 3]);
+            buf.append(ALPHABET[b0 << 2 & 0x1c | b1 >> 6]);
+            if (len <= i + 1) break;
+            buf.append(ALPHABET[b1 >> 1 & 0x1f]);
+            buf.append(ALPHABET[b1 << 4 & 0x10 | b2 >> 4]);
+            if (len <= i + 2) break;
+            buf.append(ALPHABET[b2 << 1 & 0x1e | b3 >> 7]);
+            if (len <= i + 3) break;
+            buf.append(ALPHABET[b3 >> 2 & 0x1f]);
+            buf.append(ALPHABET[b3 << 3 & 0x1c | b4 >> 5]);
+            if (len <= i + 4) break;
+            buf.append(ALPHABET[b4 & 0x1f]);
+        }
+
+        buf.append(ENCODE_PADDING[len % 5]);
+        return buf.toString();
+    }
+
+    public static boolean isValid(String encoded) {
+        return encoded.matches("[A-Z2-7=]*");
```

Comment:
This allows = anywhere in the value. A more correct approach would be to allow only trailing ='s.

Sometimes users input values with spaces, which should be ignored. Not sure if we want to handle that here as well, or require the caller to strip any whitespace characters ahead or calling decode.

Reply:
I am updating this with following: `isValid()` method only allows the base32 alphabet characters (clients have to uppercase and remove any spaces etc) with the padding character only being valid at the end to close the group.